zypper_repository: preserve attributes enabled, autorefresh and gpgcheck if provided from the task parameters

TobiasZeuch181 opened the pull request:
… task parameters

Currently, the values are overridden with the data found on the file system Instead, these values should only be updated, if they were not provided in the parameters. That way, repositories on the system could be enabled/disabled and the flags could be updated again

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
The code currently looks up the repo-file on the local disk and if it exists, it loads the optional attributes ``enabled``, ``autorefresh`` and ``gpgcheck`` and these values are used to fill up the repository-object that is later written to disk.
Unfortunately, these values are overriding the parameters, if they were provided from the task in ansible - which makes it impossible to modify the values for a repository that is already registered on the system.
So I changed the code, to:
1. set default values into
2. override with existing values from file
3. override with attributes from the ansible task configuration, if provided

Fixes #8783

##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- Bugfix Pull Request

##### COMPONENT NAME
zypper_repository


##### ADDITIONAL INFORMATION
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
1. Register a repostory, e.g. via ansible
2. use ansible to modify any of the parameters. The task will not change them

TobiasZeuch181 committed:
Don't override enabled, autorefresh and gpgcheck if provided from the parameters

Currently, the values are overridden with the data found on the file system
Instead, these values should only be updated, if they were not provided in the parameters.
That way, repositories on the system could be enabled/disabled and the flags could be updated again

TobiasZeuch181 committed:
Adding code to set default values

if the values are not provided from the task-input nor read from an existing file

felixfontein commented on plugins/modules/zypper_repository.py:
```diff
@@ -357,18 +357,21 @@ def main():
+        if module.params['autorefresh']:
+            repodata['autorefresh'] = '1'
+        else:
+            repodata['autorefresh'] = '0'
```
`module.params` always contains **all possible module options**, so `'xxx' in module.params` will always evaluate to `True` (if `xxx` is a valid module option). Also these three options have defaults, so even if the user did not provide them, they will always be set to their default value in `module.params`.

If you want to test whether the values have been provided by the user or not, you need to remove the defaults, and check for `None` instead (like `if module.params['autorefresh'] is not None:`).

TobiasZeuch181 replied:
Interesting. I thought, I had tested it locally and adding the condition works. I'll give this another try and I'll also try to add some tests

TobiasZeuch181 replied:
On the other hand I think this might break the logic where the plugin compares the configuration passed via module.params against what is found on the file system - not passing a value might result in the code detecting a change in configuration....

Also: is this the default behaviour in ansible-modules: not passing a parameter that has a default configured will change an existing configuration that was configured against the default value? That's something, I wouldn't expect

felixfontein replied:
> Also: is this the default behaviour in ansible-modules: not passing a parameter that has a default configured will change an existing configuration that was configured against the default value? That's something, I wouldn't expect

That depends on whether options have a default value or not. Having a default for options that are used for idempotency checks have this side-effect, which is usually not what the user wants. That's why it's best to only have default values for options that are not used for idempotency check, but never have defaults for options that are used for idempotency checks. (Unless there are very good reasons for a specific option; which can happen, but is not very likely.)

To fix this, it's usually best to remove the default values, and change the code so that if the user does not provide the option, the old default value is used when something is created.

Also please note that you are changing the behavior of the module, as I understand it, which is a breaking change. You need to control the behavior by a new option so that existing tasks **will behave exactly as before**.

russoz replied:
> `module.params` always contains **all possible module options**, so `'xxx' in module.params` will always evaluate to `True` (if `xxx` is a valid module option). Also these three options have defaults, so even if the user did not provide them, they will always be set to their default value in `module.params`.

I beg to differ, @felixfontein . If the value of the parameter is `None`, then that stmt is equivalent to `if None:`, which evaluates to `False`. He's not checking whether the parameter exists, he is using the parameter value, and `None` is... `False`-ish.

felixfontein replied:
The statement `if 'xxx' in module.params:` is **not** equivalent to `if None` if `module.params['xxx']` is `None`.

russoz replied:
You're right of course. I was mislead by the fact that the lines of code selected for this comment do not include that part. Re-reading I see I missed some bits, and I have even submitted a suggestion fixing that up.

felixfontein commented on changelogs/fragments/8783-preserve-parameters-provided-from-task-in-zypper-module.yaml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - keycloak_client - fix TypeError when sanitizing the ``saml.signing.private.key`` attribute in the module's diff or state output. The ``sanitize_cr`` function expected a dict where in some cases a list might occur (https://github.com/ansible-collections/community.general/pull/8403).
```
This message is about another module and doesn't make sense for this one.

TobiasZeuch181 replied:
Right, sorry, I had edited the changes but not saved them...

felixfontein reviewed (commented):
Thanks for your contribution! I've added some first comments.

TobiasZeuch181 committed:
Reordering how the boolean parameters are filled

I think, this is the right approach:

# fill the boolean entries with default
# override the values with what is found in the repo-file (remote or locally on the file)
# finally override the values with what is provided in the module-params

then we can check if the data has changed compared to the local data

TobiasZeuch181 committed:
Updating the content of the changelog file

I had copied and edited the data but not saved the changes

TobiasZeuch181 committed:
Adding tests for modifying the repository

adding tests:
* modifying a repository should work (disabling)
* modifying a different parameter doesn't change configuation that was not specified - although it has a different default-value, than what is currenctly configured

felixfontein commented on changelogs/fragments/8783-preserve-parameters-provided-from-task-in-zypper-module.yaml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - zypper-module - only read attributes ``enabled``, ``disable_gpg_check`` and ``autorefresh`` from local repo files, if the parameters weren't already provided by the task parameters
```
```suggestion
  - zypper-module - only read attributes ``enabled``, ``disable_gpg_check`` and ``autorefresh`` from local repo files, if the parameters were not already provided by the task parameters (https://github.com/ansible-collections/community.general/pull/9108).
```

russoz replied:
Hi @TobiasZeuch181 please add this modification as well.

russoz commented on plugins/modules/zypper_repository.py:
```diff
@@ -393,6 +380,15 @@ def exit_unchanged():
         if not alias and state == "present":
             module.fail_json(msg='Name required when adding non-repo files.')
 
+
```
This extra line is tripping the PEP8 sanity check.
```suggestion
```

TobiasZeuch181 committed:
Remove extra line that breaks the PEP8 sanity check

Co-authored-by: Alexei Znamensky <103110+russoz@users.noreply.github.com>

TobiasZeuch181 committed:
remove defaults and replace logic for checking defined parameters

russoz commented:
Other than the small detail in the changelog fragment, LGTM

russoz commented on plugins/modules/zypper_repository.py:
```diff
@@ -443,6 +438,24 @@ def exit_unchanged():
 
+    # override boolean parameters in repodata with provided entries in module.params
+    # in the language that zypper lr -x provides for easier comparison
+    if 'enabled' in module.params:
+        if module.params['enabled']:
+            repodata['enabled'] = '1'
+        else:
+            repodata['enabled'] = '0'
+    if 'disable_gpg_check' in module.params:
+        if module.params['disable_gpg_check']:
+            repodata['gpgcheck'] = '0'
+        else:
+            repodata['gpgcheck'] = '1'
+    if 'autorefresh' in module.params:
+        if module.params['autorefresh']:
+            repodata['autorefresh'] = '1'
+        else:
+            repodata['autorefresh'] = '0'
```
This check if the key exists, however, is redundant. They will always exist. Please simplify this to:
```suggestion
    repodata['enabled'] = '1' if module.params['enabled'] else '0'
    repodata['gpgcheck'] = '0' if module.params['disable_gpg_check'] else '1'
    repodata['autorefresh'] = '1' if module.params['autorefresh'] else '0'
```

TobiasZeuch181 replied:
I think, it would be more intuitive to remove the default values and then have the additional check for the flags in module.params. 
Then you could do the following:

  - name: Disable the repository
    zypper_repository:
      repo: repo_path
      enabled: false
  - name: Disable the gpg-check
    zypper_repository:
      repo: repo_path
      disable_gpg_check: false

and the repository would still be disabled. Currently - with the default values - the second task would enable it again.

I understand though, that this changes the behaviour of the module (against what was previously declared in the documentation, because actually, the update functionality is broken completely when all data is overridden with what is found on the file system).
On the other hand, I can't picture many situations, where you need two different modifications on the same module in two steps. So I am ok with what you suggested

felixfontein commented:
@TobiasZeuch181 ping

needs_info

TobiasZeuch181 commented:
Sorry for the delay. I am slowly progressing with setting up my test environment for the improvement. It seems, that it still doesn't help to "simply" disable a repository that is already installed.
I linked the file into a different ansible-repository and simply call the module via molecule there. Debugging would be nice, not sure if Imanage to set that up though

TobiasZeuch181 committed:
Simplify the check for flags

and add the defaults again, which I had removed in the previous step, to maintain the behaviour according to the previous documentation

TobiasZeuch181 committed:
Update changelogs/fragments/8783-preserve-parameters-provided-from-task-in-zypper-module.yaml

Co-authored-by: Felix Fontein <felix@fontein.de>

TobiasZeuch181 committed:
Use correct default-value for autorefresh

felixfontein commented on plugins/modules/zypper_repository.py:
```diff
@@ -443,6 +438,12 @@ def exit_unchanged():
+    # in the language that zypper lr -x provides for easier comparison
+    repodata['enabled'] = '1' if module.params['enabled'] else '0'
+    repodata['gpgcheck'] = '0' if module.params['disable_gpg_check'] else '1'
+    repodata['autorefresh'] = '1' if module.params['autorefresh'] else '0'
```
The above three lines are indiscriminantly overwriting the values from the repo file (and the defaults set in the code above) with the values provided by the user (resp. the module defaults).

(Also this now overrides the values from the repo, while before the repo values overrode the values provided by the user. I guess both are wrong, but the current way is a breaking change and the old way has been the case for a longer time, so the module should stick to it unless the user explicitly tells the module to not do so.)

TobiasZeuch181 replied:
Hi Felix,

sorry, I don't get why do you say that the current implementation is probably wrong. Updating the value in the repo-file with the value provided by the user is the purpose of the module. And updating with the module-default is something that I can't prevent, unless I remove the default-value from the parameter-definition, right?

About this being a breaking change: I agree, but the current implementation makes it impossible to update any of these fields. Calling the module with an existing repo-path will succeed but have no impact on the repository configuration at all. 
IF someone had such a task in his playbook that had no effect until now then the playbook would behave differently though with my change. 

So what would you suggest here? You said before that I would need to introduce a new parameter to activate the new behaviour, would that be the way to go?

felixfontein replied:
See https://github.com/ansible-collections/community.general/pull/9108#discussion_r1864266288. You need to remove the defaults and handle them in code.

felixfontein commented:
Please note that this PR now has a conflict since we reformatted all code in the collection (see #10999).

felixfontein commented:
Ping @TobiasZeuch181 

needs_info
